<a href="https://colab.research.google.com/github/joselmuniz0/gestion-con-python/blob/main/TPI_DataAnalytics_Etapa1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

El trabajo está muy bien desarrollado, organizado y documentado, mostrando un nivel más avanzado en comparación con otros entregables. Se destacan la correcta estructuración del notebook, el uso de funciones reutilizables para el análisis exploratorio y calidad de datos, la explicación técnica de cada proceso y la buena documentación de objetivos, conclusiones y pasos realizados. También es positivo el uso de copias de DataFrames para no modificar los datos originales y la justificación clara de las estructuras de datos elegidas. Como mejora, algunas secciones podrían resumirse un poco para evitar demasiada extensión y repetición de explicaciones, especialmente en comentarios y descripciones teóricas. Además, en la integración de datos sería recomendable controlar los duplicados en marketing antes del merge para evitar multiplicar registros. En general, es un trabajo muy completo, prolijo y profesional que cumple ampliamente con los objetivos de la Etapa 1.

# Trabajo Práctico Integrador — Data Analytics con Python
## Etapa 1: Recopilación y Preparación de Datos

| | |
|---|---|
| **Autor** | Muñiz Jose |
| **Comisión** | 26112 |
| **Fecha** | 08-05-2026
| Análisis de Ventas y Efectividad de Campañas de Marketing |


### Datasets utilizados
- **`clientes.csv`** — datos demográficos de los clientes (567 registros).
- **`ventas.csv`** — transacciones individuales de venta (3.035 registros).
- **`marketing.csv`** — campañas de marketing realizadas por producto (90 registros).

### Actividades
1. Carga de los 3 datasets como DataFrames.
2. Cálculo de ventas mensuales con variables y operadores.
3. Almacenamiento de ventas con estructuras de datos (lista de diccionarios).
4. Análisis exploratorio inicial con Pandas.
5. Identificación de valores nulos y duplicados (calidad de datos).

## 1. Objetivos

**Objetivo general:** preparar los datasets de ventas, clientes y campañas de marketing para su análisis posterior, documentando el estado inicial de los datos.

**Objetivos específicos:**
- Cargar correctamente los 3 archivos CSV como `pandas.DataFrame`.
- Implementar un cálculo de ventas mensuales con Python y con Pandas.
- Modelar registros de ventas mediante una estructura de datos apropiada y justificar la elección.
- Realizar un análisis exploratorio inicial sistemático y reproducible sobre los 3 datasets.
- Detectar y reportar nulos y duplicados


## 2. Setup del entorno

Esta sección prepara el entorno: carga las librerías, monta Google Drive y centraliza las rutas en constantes.

In [1]:
# 2.1 — Imports estándar
# pandas: manipulación tabular de datos
# numpy: operaciones numéricas (lo usa pandas internamente)
# manejo de rutas del sistema de archivos
import os
import pandas as pd
import numpy as np

# Configuración de visualización para que el EDA sea legible en pantalla
pd.set_option('display.max_columns', 50)            # mostrar hasta 50 columnas sin recortar
pd.set_option('display.width', 160)                 # ancho de impresión cómodo
pd.set_option('display.float_format', '{:,.2f}'.format)  # formato numérico con separador de miles

print('Pandas:', pd.__version__)
print('NumPy :', np.__version__)

Pandas: 2.2.2
NumPy : 2.0.2


In [2]:
# 2.2 — Montaje de Google Drive
# Al ejecutar esta celda, Colab abre un pop-up pidiendo autorización.
# Una vez autorizado, "Mi unidad" queda accesible en /content/drive/MyDrive/
from google.colab import drive
drive.mount('/content/drive/')

Mounted at /content/drive/


In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
import os

# 2.3 — Definición de rutas del proyecto
# Centralizamos todas las rutas en constantes. Si la carpeta cambia de nombre o
# se renombra algún archivo, basta con editar esta celda — no hay que tocar el resto.

# Original path (commented out as it caused the error)
# RUTA_PROYECTO = '/content/drive/MyDrive/Muñiz Jose - Comisión 26112 - TPI Data Analytics'

# Updated path: Assuming data files are directly in MyDrive based on kernel context
RUTA_PROYECTO = '/content/drive/MyDrive/Muniz Jose - Comisión 26112 - TPI Data Analytics'
RUTA_DATA     = '/content/drive/MyDrive/Muniz Jose - Comisión 26112 - TPI Data Analytics' # Data files are in the same location as the project root in this case

PATH_CLIENTES  = os.path.join(RUTA_DATA, 'clientes.csv')
PATH_VENTAS    = os.path.join(RUTA_DATA, 'ventas.csv')
PATH_MARKETING = os.path.join(RUTA_DATA, 'marketing.csv')

# Validación temprana: si algo no está, el assert falla acá con mensaje claro
# en lugar de hacerlo tres celdas más abajo con un error opaco.
assert os.path.isdir(RUTA_PROYECTO), f'No existe la carpeta del proyecto: {RUTA_PROYECTO}'
assert os.path.isdir(RUTA_DATA),     f'No existe la carpeta data: {RUTA_DATA}'
for path in [PATH_CLIENTES, PATH_VENTAS, PATH_MARKETING]:
    assert os.path.isfile(path), f'No se encontró el archivo: {path}'

print('Rutas validadas correctamente.')
print('\nContenido de la carpeta data/:')
for f in sorted(os.listdir(RUTA_DATA)):
    print(' -', f)

Rutas validadas correctamente.

Contenido de la carpeta data/:
 - Datos con python
 - TPI_DataAnalytics_Etapa1.ipynb
 - clientes.csv
 - marketing.csv
 - ventas.csv
 - 📂 Pre-Entrega de Proyecto.pdf


## 3. Actividad 1 — Carga de los 3 datasets como DataFrames

**se leen los 3 CSV y se cargan como objetos `pandas.DataFrame`.





In [5]:
# 3.1 — Carga del dataset de Clientes
df_clientes = pd.read_csv(PATH_CLIENTES, sep=',', encoding='utf-8')

print(f'df_clientes: {df_clientes.shape[0]} filas, {df_clientes.shape[1]} columnas')
df_clientes.head()


df_clientes: 567 filas, 5 columnas


,id_cliente,nombre,edad,ciudad,ingresos
0,1,Aloysia Screase,44,Mar del Plata,"42,294.68"
1,2,Kristina Scaplehorn,25,Posadas,"24,735.04"
2,3,Filip Castagne,50,Resistencia,"35,744.85"
3,4,Liuka Luard,39,Bahía Blanca,"27,647.96"
4,5,Dore Cockshtt,28,Rosario,"28,245.65"


In [6]:
# 3.2 — Carga del dataset de Ventas
df_ventas = pd.read_csv(PATH_VENTAS, sep=',', encoding='utf-8')

print(f'df_ventas: {df_ventas.shape[0]} filas, {df_ventas.shape[1]} columnas')
df_ventas.head()


df_ventas: 3035 filas, 6 columnas


,id_venta,producto,precio,cantidad,fecha_venta,categoria
0,792,Cuadro decorativo,$69.94,5.00,02/01/2024,Decoración
1,811,Lámpara de mesa,$105.10,5.00,02/01/2024,Decoración
2,1156,Secadora,$97.96,3.00,02/01/2024,Electrodomésticos
3,1372,Heladera,$114.35,8.00,02/01/2024,Electrodomésticos
4,1546,Secadora,$106.21,4.00,02/01/2024,Electrodomésticos


In [7]:
# 3.3 — Carga del dataset de Marketing
df_marketing = pd.read_csv(PATH_MARKETING, sep=',', encoding='utf-8')

print(f'df_marketing: {df_marketing.shape[0]} filas, {df_marketing.shape[1]} columnas')
df_marketing.head()


df_marketing: 90 filas, 6 columnas


,id_campanha,producto,canal,costo,fecha_inicio,fecha_fin
0,74,Adorno de pared,TV,4.81,20/03/2024,03/05/2024
1,12,Tablet,RRSS,3.40,26/03/2024,13/05/2024
2,32,Lámpara de mesa,Email,5.54,28/03/2024,20/04/2024
3,21,Smartphone,RRSS,6.37,29/03/2024,16/05/2024
4,58,Alfombra,Email,4.25,31/03/2024,05/05/2024


### 3.4 — Verificación rápida de la carga

Resumen consolidado para confirmar que los 3 datasets quedaron cargados correctamente.

In [8]:
# 3.4 — Tabla resumen de los 3 datasets
resumen_carga = pd.DataFrame({
    'dataset': ['clientes', 'ventas', 'marketing'],
    'filas': [df_clientes.shape[0], df_ventas.shape[0], df_marketing.shape[0]],
    'columnas': [df_clientes.shape[1], df_ventas.shape[1], df_marketing.shape[1]],
    'memoria_KB': [
        round(df_clientes.memory_usage(deep=True).sum() / 1024, 2),
        round(df_ventas.memory_usage(deep=True).sum() / 1024, 2),
        round(df_marketing.memory_usage(deep=True).sum() / 1024, 2),
    ],
})
resumen_carga


,dataset,filas,columnas,memoria_KB
0,clientes,567,5,84.81
1,ventas,3035,6,868.64
2,marketing,90,6,22.62


In [9]:
# 4.1 — Cálculo de ventas mensuales con Python

# Datos sintéticos de ejemplo (formato similar al dataset real)
ventas_demo = [
    {'fecha': '02/01/2024', 'producto': 'Cuadro decorativo', 'precio': 69.94, 'cantidad': 5},
    {'fecha': '15/01/2024', 'producto': 'Lámpara de mesa',   'precio': 105.10, 'cantidad': 3},
    {'fecha': '03/02/2024', 'producto': 'Heladera',          'precio': 850.00, 'cantidad': 2},
    {'fecha': '20/02/2024', 'producto': 'Tablet',            'precio': 320.00, 'cantidad': 4},
    {'fecha': '10/03/2024', 'producto': 'Smartphone',        'precio': 600.00, 'cantidad': 6},
]

# Acumulador: clave = mes (YYYY-MM), valor = importe total facturado en ese mes
ventas_por_mes = {}

for registro in ventas_demo:
    # La fecha viene como 'dd/mm/yyyy'. Extraemos mes y año por slicing.
    dia, mes, anio = registro['fecha'].split('/')   # ['02', '01', '2024']
    clave_mes = f'{anio}-{mes}'                     # '2024-01' (formato ordenable)

    # Operador aritmético: precio unitario * cantidad = importe de la línea
    importe = registro['precio'] * registro['cantidad']

    # .get(clave, 0) evita KeyError la primera vez que aparece un mes nuevo
    ventas_por_mes[clave_mes] = ventas_por_mes.get(clave_mes, 0) + importe

# Imprimimos ordenado cronológicamente (sorted ordena las claves alfabéticamente,
# y como usamos formato YYYY-MM eso coincide con el orden cronológico).
print('Ventas mensuales (versión didáctica con Python):')
print('-' * 40)
for mes in sorted(ventas_por_mes):
    print(f'  {mes}:  $ {ventas_por_mes[mes]:>12,.2f}')


Ventas mensuales (versión didáctica con Python):
----------------------------------------
  2024-01:  $       665.00
  2024-02:  $     2,980.00
  2024-03:  $     3,600.00


### 4.2 — Versión con Pandas (sobre los datos reales)

Resolvemos el mismo cálculo sobre `df_ventas` usando vectorización. Es órdenes de magnitud más eficiente y es lo que se haría en un entorno profesional.

**Pasos:**
1. Crear una **copia local** del DataFrame (no tocamos el original durante el diagnóstico).
2. Limpiar localmente la columna `precio` para poder operar numéricamente con ella.
3. Convertir `fecha_venta` a `datetime`.
4. Calcular el importe línea por línea.
5. Agrupar por mes y sumar.

In [10]:
# 4.2 — Cálculo de ventas mensuales con Pandas

# 1) Copia local: cualquier transformación que hagamos acá NO afecta a df_ventas.
#    Esto es clave en la Etapa 1: estamos diagnosticando, no limpiando.
ventas_calc = df_ventas.copy()

# 2) Limpieza local de la columna 'precio':
#    - .str.replace('$', '', regex=False) saca el símbolo monetario
#    - .astype(float) convierte el resultado a número
#    - errors='coerce' en pd.to_numeric: si algo no se puede convertir, queda NaN
ventas_calc['precio'] = pd.to_numeric(
    ventas_calc['precio'].str.replace('$', '', regex=False),
    errors='coerce'
)

# 3) Conversión de fecha: 'dd/mm/yyyy' -> datetime
#    dayfirst=True le indica a pandas el orden día-mes-año típico de Argentina
ventas_calc['fecha_venta'] = pd.to_datetime(
    ventas_calc['fecha_venta'], dayfirst=True, errors='coerce'
)

# 4) Cálculo del importe: precio * cantidad. Pandas vectoriza esto automáticamente.
ventas_calc['importe'] = ventas_calc['precio'] * ventas_calc['cantidad']

# 5) Agrupamiento por mes:
#    .dt.to_period('M') convierte cada fecha en un período mensual ('2024-01').
#    A diferencia de strftime(), preserva el orden cronológico nativo.
ventas_mensuales = (
    ventas_calc
    .groupby(ventas_calc['fecha_venta'].dt.to_period('M'))['importe']
    .sum()
    .sort_index()
)

print('Ventas mensuales reales (Pandas):')
print('-' * 40)
print(ventas_mensuales.apply(lambda x: f'$ {x:>14,.2f}').to_string())
print('-' * 40)
print(f'TOTAL del período: $ {ventas_mensuales.sum():,.2f}')


Ventas mensuales reales (Pandas):
----------------------------------------
fecha_venta
2024-01    $     129,604.99
2024-02    $     118,672.44
2024-03    $     136,779.15
2024-04    $     144,380.10
2024-05    $     143,727.25
2024-06    $     108,480.17
2024-07    $     116,229.97
2024-08    $     119,680.15
2024-09    $     115,787.85
2024-10    $     112,117.13
2024-11    $     119,951.79
2024-12    $     117,631.94
Freq: M
----------------------------------------
TOTAL del período: $ 1,483,042.93


**Lectura del resultado:** el dataset cubre el período de febrero a octubre de 2024. Las dos filas con nulos en `precio` o `cantidad` se excluyen automáticamente del cálculo (al ser `NaN`, el producto da `NaN` y `sum()` lo ignora).

## 5. Actividad 3 — Estructuras de datos para almacenar ventas

Modelado de registros de ventas (producto, precio, cantidad) y argumentacion de la elección de estructura.

### Análisis de alternativas

| Modelo | Registro | Pros | Contras |
|---|---|---|---|
| Lista de listas | `['Heladera', 850.0, 2]` | Mínimo overhead, simple | Acceso por índice numérico (frágil si cambia el orden) |
| Lista de tuplas | `('Heladera', 850.0, 2)` | Inmutable: protege la integridad de cada venta | Sigue siendo posicional, no por nombre |
| **Lista de diccionarios** | `{'producto':'Heladera', 'precio':850.0, 'cantidad':2}` | **Acceso por nombre, autodescriptivo, convertible directo a DataFrame** | Más uso de memoria |
| Diccionario por producto | `{'Heladera': {'precio':850, 'unid':2}}` | Búsqueda O(1) por producto | Pierde detalle de ventas individuales (colapsa los duplicados) |

### Decisión: lista de diccionarios

**Razones técnicas:**
1. Cada venta tiene **campos heterogéneos con nombre** → el diccionario es la representación natural.
2. La **lista preserva el orden cronológico** de las ventas y admite duplicados legítimos (dos ventas del mismo producto el mismo día son válidas y no deben colapsarse).
3. Es **directamente convertible a `pandas.DataFrame`** con `pd.DataFrame(lista)` — esto cierra el ciclo con el resto del análisis.
4. Cuando se lee `v['precio']` se entiende inmediatamente qué se está accediendo, mientras que `v[1]` requiere recordar el orden de las columnas.

In [11]:
# 5 — Almacenamiento de ventas con lista de diccionarios

# Subconjunto representativo del dataset real (productos de las 3 categorías)
ventas = [
    {'producto': 'Cuadro decorativo', 'precio':  69.94, 'cantidad': 5},
    {'producto': 'Lámpara de mesa',   'precio': 105.10, 'cantidad': 3},
    {'producto': 'Heladera',          'precio': 850.00, 'cantidad': 2},
    {'producto': 'Tablet',            'precio': 320.00, 'cantidad': 4},
    {'producto': 'Smartphone',        'precio': 600.00, 'cantidad': 6},
    {'producto': 'Heladera',          'precio': 850.00, 'cantidad': 1},  # duplicado legítimo
]

# 1) Iteración explícita: imprimir cada venta con su importe
print('Detalle línea por línea:')
print('-' * 60)
for v in ventas:
    importe = v['precio'] * v['cantidad']
    print(f"  {v['producto']:<20}  ${v['precio']:>7.2f} x {v['cantidad']}  =  ${importe:>10,.2f}")

# 2) Total general usando una generator expression dentro de sum()
#    Esto es más eficiente que crear una lista intermedia con [v['precio']*... for v in ventas]
total = sum(v['precio'] * v['cantidad'] for v in ventas)
print('-' * 60)
print(f'Total facturado: $ {total:,.2f}\n')

# 3) Total por producto: patrón acumulador con diccionario
totales_por_producto = {}
for v in ventas:
    p = v['producto']
    totales_por_producto[p] = totales_por_producto.get(p, 0) + v['precio'] * v['cantidad']

print('Total por producto:')
print('-' * 40)
for p, t in sorted(totales_por_producto.items(), key=lambda x: -x[1]):  # ordenado desc
    print(f'  {p:<20}  $ {t:>10,.2f}')

# 4) Conversión inmediata a DataFrame: puente con el resto del análisis
df_ventas_demo = pd.DataFrame(ventas)
print('\nMismo conjunto como DataFrame:')
df_ventas_demo


Detalle línea por línea:
------------------------------------------------------------
  Cuadro decorativo     $  69.94 x 5  =  $    349.70
  Lámpara de mesa       $ 105.10 x 3  =  $    315.30
  Heladera              $ 850.00 x 2  =  $  1,700.00
  Tablet                $ 320.00 x 4  =  $  1,280.00
  Smartphone            $ 600.00 x 6  =  $  3,600.00
  Heladera              $ 850.00 x 1  =  $    850.00
------------------------------------------------------------
Total facturado: $ 8,095.00

Total por producto:
----------------------------------------
  Smartphone            $   3,600.00
  Heladera              $   2,550.00
  Tablet                $   1,280.00
  Cuadro decorativo     $     349.70
  Lámpara de mesa       $     315.30

Mismo conjunto como DataFrame:


,producto,precio,cantidad
0,Cuadro decorativo,69.94,5
1,Lámpara de mesa,105.10,3
2,Heladera,850.00,2
3,Tablet,320.00,4
4,Smartphone,600.00,6
5,Heladera,850.00,1


## 6. Actividad 4 — Análisis exploratorio inicial (EDA)

**Qué se hace:** describir cada dataset *antes* de cualquier transformación, respondiendo:
- ¿Cuántos registros y columnas tiene?
- ¿Qué tipo tiene cada columna?
- ¿Cómo se distribuyen las variables numéricas?
- ¿Qué cardinalidad tienen las columnas categóricas?

**Por qué se hace así:** sin entender el dato, cualquier limpieza posterior sería a ciegas. El EDA es el insumo que nos permite decidir, en la Etapa 2, **qué** hay que limpiar y **cómo**.

Encapsulamos la exploración en una función reutilizable para garantizar que los 3 datasets se traten con el mismo criterio sistemático.

In [12]:
# 6.1 — Función reutilizable de EDA

def explorar_dataframe(df: pd.DataFrame, nombre: str) -> None:
    '''
    Imprime un resumen exploratorio estándar de un DataFrame.

    Parameters
    ----------
    df : pd.DataFrame
        DataFrame a explorar.
    nombre : str
        Nombre descriptivo del dataset (aparece en los encabezados).
    '''
    sep = '=' * 70
    print(sep)
    print(f'  EDA — {nombre}')
    print(sep)

    # 1. Dimensiones del dataset
    print(f'\nForma: {df.shape[0]} filas x {df.shape[1]} columnas')

    # 2. Tipos de datos y uso de memoria
    print('\nTipos de datos y memoria:')
    df.info(memory_usage='deep')

    # 3. Primeras filas
    print('\nPrimeras 5 filas:')
    print(df.head())

    # 4. Últimas filas (útil para detectar registros incompletos al final)
    print('\nÚltimas 5 filas:')
    print(df.tail())

    # 5. Estadísticos descriptivos de variables numéricas
    print('\nEstadísticos numéricos:')
    print(df.describe())

    # 6. Estadísticos de variables categóricas (si existen)
    cols_obj = df.select_dtypes(include=['object', 'category']).columns
    if len(cols_obj) > 0:
        print('\nEstadísticos de columnas categóricas:')
        print(df[cols_obj].describe())

    # 7. Cardinalidad: cuántos valores únicos tiene cada columna
    print('\nValores únicos por columna:')
    print(df.nunique().sort_values(ascending=False))

    print('\n' + sep + '\n')


In [13]:
# 6.2 — EDA del dataset Clientes
explorar_dataframe(df_clientes, 'Clientes')


  EDA — Clientes

Forma: 567 filas x 5 columnas

Tipos de datos y memoria:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 567 entries, 0 to 566
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id_cliente  567 non-null    int64  
 1   nombre      567 non-null    object 
 2   edad        567 non-null    int64  
 3   ciudad      567 non-null    object 
 4   ingresos    567 non-null    float64
dtypes: float64(1), int64(2), object(2)
memory usage: 84.8 KB

Primeras 5 filas:
   id_cliente               nombre  edad         ciudad  ingresos
0           1      Aloysia Screase    44  Mar del Plata 42,294.68
1           2  Kristina Scaplehorn    25        Posadas 24,735.04
2           3       Filip Castagne    50    Resistencia 35,744.85
3           4          Liuka Luard    39   Bahía Blanca 27,647.96
4           5        Dore Cockshtt    28        Rosario 28,245.65

Últimas 5 filas:
     id_cliente           nombre  edad 

In [14]:
# 6.3 — EDA del dataset Ventas
explorar_dataframe(df_ventas, 'Ventas')


  EDA — Ventas

Forma: 3035 filas x 6 columnas

Tipos de datos y memoria:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 868.6 KB

Primeras 5 filas:
   id_venta           producto   precio  cantidad fecha_venta          categoria
0       792  Cuadro decorativo   $69.94      5.00  02/01/2024         Decoración
1       811    Lámpara de mesa  $105.10      5.00  02/01/2024         Decoración
2      1156           Secadora   $97.96      3.00  02/01/2024  Electrodomésticos
3      1372           Heladera  $114.35      8.00  02/01/2024  Electrodoméstic

In [15]:
# 6.4 — EDA del dataset Marketing
explorar_dataframe(df_marketing, 'Marketing')


  EDA — Marketing

Forma: 90 filas x 6 columnas

Tipos de datos y memoria:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id_campanha   90 non-null     int64  
 1   producto      90 non-null     object 
 2   canal         90 non-null     object 
 3   costo         90 non-null     float64
 4   fecha_inicio  90 non-null     object 
 5   fecha_fin     90 non-null     object 
dtypes: float64(1), int64(1), object(4)
memory usage: 22.6 KB

Primeras 5 filas:
   id_campanha         producto  canal  costo fecha_inicio   fecha_fin
0           74  Adorno de pared     TV   4.81   20/03/2024  03/05/2024
1           12           Tablet   RRSS   3.40   26/03/2024  13/05/2024
2           32  Lámpara de mesa  Email   5.54   28/03/2024  20/04/2024
3           21       Smartphone   RRSS   6.37   29/03/2024  16/05/2024
4           58         Alfombra  Email   4.

### 6.5 — Lectura de los resultados del EDA

**Dataset Clientes (567 registros, 5 columnas):**
- Columnas: `id_cliente`, `nombre`, `edad`, `ciudad`, `ingresos`.
- Tipos correctos: `id_cliente` y `edad` son enteros; `ingresos` es float; `nombre` y `ciudad` son texto.
- 12 ciudades argentinas distintas.
- A primera vista, no hay tipos sospechosos.

**Dataset Ventas (3.035 registros, 6 columnas):**
- Columnas: `id_venta`, `producto`, `precio`, `cantidad`, `fecha_venta`, `categoria`.
- **`precio` aparece como `object` (string)**, no como número. Causa: viene con el símbolo `$` adelante (ej. `'$69.94'`). Esto va a requerir limpieza en la Etapa 2.
- **`fecha_venta` también es `object`**, no `datetime`. Va a haber que convertirla.
- 30 productos distintos repartidos en 3 categorías: Decoración, Electrodomésticos, Electrónica.
- Período cubierto: enero a diciembre de 2024 (año completo).

**Dataset Marketing (90 registros, 6 columnas):**
- Columnas: `id_campanha`, `producto`, `canal`, `costo`, `fecha_inicio`, `fecha_fin`.
- **`fecha_inicio` y `fecha_fin` son `object`** — habrá que convertirlas a `datetime`.
- 3 canales de marketing: TV, RRSS, Email.
- Los 30 productos coinciden con los productos del dataset de Ventas (excelente: permite cruzar ambos datasets por `producto` en la Etapa 2).


## 7. Actividad 5 — Calidad de datos: nulos y duplicados

Se identifican valores nulos y registros duplicados, se emite un informe consolidado del estado inicial.



In [16]:
# 7.1 — Función de informe de nulos por columna (con porcentaje)

def informe_nulos(df: pd.DataFrame, nombre: str) -> pd.DataFrame:
    '''
    Devuelve un DataFrame con cantidad y porcentaje de nulos por columna.

    Reportar cantidad y porcentaje permite priorizar el tratamiento:
    - <5% de nulos → suele admitir imputación o eliminación de filas.
    - 5–30%       → analizar con cuidado el impacto.
    - >30%        → considerar descartar la columna.
    '''
    cantidad   = df.isnull().sum()
    porcentaje = (cantidad / len(df) * 100).round(2)
    informe = (
        pd.DataFrame({
            'columna': cantidad.index,
            'nulos': cantidad.values,
            '% nulos': porcentaje.values,
        })
        .sort_values('nulos', ascending=False)
        .reset_index(drop=True)
    )
    print(f'\n--- Nulos en {nombre} ---')
    print(f'Total de nulos en el dataset: {int(df.isnull().sum().sum())}')
    print(informe.to_string(index=False))
    return informe

nulos_clientes  = informe_nulos(df_clientes,  'Clientes')
nulos_ventas    = informe_nulos(df_ventas,    'Ventas')
nulos_marketing = informe_nulos(df_marketing, 'Marketing')



--- Nulos en Clientes ---
Total de nulos en el dataset: 0
   columna  nulos  % nulos
id_cliente      0     0.00
    nombre      0     0.00
      edad      0     0.00
    ciudad      0     0.00
  ingresos      0     0.00

--- Nulos en Ventas ---
Total de nulos en el dataset: 4
    columna  nulos  % nulos
   cantidad      2     0.07
     precio      2     0.07
   producto      0     0.00
   id_venta      0     0.00
fecha_venta      0     0.00
  categoria      0     0.00

--- Nulos en Marketing ---
Total de nulos en el dataset: 0
     columna  nulos  % nulos
 id_campanha      0     0.00
    producto      0     0.00
       canal      0     0.00
       costo      0     0.00
fecha_inicio      0     0.00
   fecha_fin      0     0.00


In [17]:
# 7.2 — Función de informe de duplicados

def informe_duplicados(df: pd.DataFrame, nombre: str, subset=None) -> pd.DataFrame:
    '''
    Reporta y devuelve los registros duplicados de un DataFrame.

    Parameters
    ----------
    subset : list[str] | None
        Columnas que definen 'duplicado'. Si es None, considera todas.

    Notas:
    - keep=False marca *todas* las apariciones del duplicado, no solo las repeticiones.
    - keep='first' (default) marcaría solo las repeticiones, ocultando la primera ocurrencia.
    '''
    mask = df.duplicated(subset=subset, keep=False)
    duplicados = df[mask].sort_values(by=subset or list(df.columns))

    print(f'\n--- Duplicados en {nombre} ---')
    print(f'Filas totales: {len(df)}')
    print(f'Filas duplicadas (todas las apariciones): {len(duplicados)}')
    print(f'Filas únicas si se eliminaran duplicados: '
          f'{len(df) - int(df.duplicated(subset=subset).sum())}')
    if subset:
        print(f'Criterio de duplicación: columnas {subset}')
    return duplicados

dup_clientes  = informe_duplicados(df_clientes,  'Clientes')
dup_ventas    = informe_duplicados(df_ventas,    'Ventas')
dup_marketing = informe_duplicados(df_marketing, 'Marketing')



--- Duplicados en Clientes ---
Filas totales: 567
Filas duplicadas (todas las apariciones): 0
Filas únicas si se eliminaran duplicados: 567

--- Duplicados en Ventas ---
Filas totales: 3035
Filas duplicadas (todas las apariciones): 70
Filas únicas si se eliminaran duplicados: 3000

--- Duplicados en Marketing ---
Filas totales: 90
Filas duplicadas (todas las apariciones): 0
Filas únicas si se eliminaran duplicados: 90


In [18]:
# 7.3 — Informe consolidado del estado inicial

def informe_inicial(df: pd.DataFrame, nombre: str) -> dict:
    '''Imprime y devuelve el informe inicial de calidad de un dataset.'''
    print(f'\n{"#"*60}')
    print(f'  INFORME INICIAL — {nombre}')
    print(f'{"#"*60}')

    total_registros = len(df)
    nulos_por_col   = df.isnull().sum()
    total_nulos     = int(nulos_por_col.sum())
    duplicados      = int(df.duplicated().sum())

    print(f'\na) Cantidad total de registros: {total_registros}')
    print(f'\nb) Valores nulos por columna:')
    if total_nulos > 0:
        print(nulos_por_col[nulos_por_col > 0].to_string())
    else:
        print('   (no hay nulos)')
    print(f'\n   Total de nulos en el dataset: {total_nulos}')
    print(f'\nc) Filas duplicadas (criterio: todas las columnas): {duplicados}')

    return {
        'dataset': nombre,
        'registros': total_registros,
        'nulos_total': total_nulos,
        'nulos_por_columna': nulos_por_col.to_dict(),
        'duplicados_total': duplicados,
    }

resumen_clientes  = informe_inicial(df_clientes,  'Clientes')
resumen_ventas    = informe_inicial(df_ventas,    'Ventas')
resumen_marketing = informe_inicial(df_marketing, 'Marketing')



############################################################
  INFORME INICIAL — Clientes
############################################################

a) Cantidad total de registros: 567

b) Valores nulos por columna:
   (no hay nulos)

   Total de nulos en el dataset: 0

c) Filas duplicadas (criterio: todas las columnas): 0

############################################################
  INFORME INICIAL — Ventas
############################################################

a) Cantidad total de registros: 3035

b) Valores nulos por columna:
precio      2
cantidad    2

   Total de nulos en el dataset: 4

c) Filas duplicadas (criterio: todas las columnas): 35

############################################################
  INFORME INICIAL — Marketing
############################################################

a) Cantidad total de registros: 90

b) Valores nulos por columna:
   (no hay nulos)

   Total de nulos en el dataset: 0

c) Filas duplicadas (criterio: todas las columnas): 0


### 7.4 — DataFrame con los registros duplicados



In [19]:
# 7.4.1 — Duplicados completos en Clientes
duplicados_clientes_df = df_clientes[df_clientes.duplicated(keep=False)]
print(f'Duplicados completos en Clientes: {len(duplicados_clientes_df)} filas')
duplicados_clientes_df


Duplicados completos en Clientes: 0 filas


,id_cliente,nombre,edad,ciudad,ingresos


In [20]:
# 7.4.2 — Duplicados completos en Ventas
duplicados_ventas_df = df_ventas[df_ventas.duplicated(keep=False)]
print(f'Duplicados completos en Ventas: {len(duplicados_ventas_df)} filas')
duplicados_ventas_df.head(20)   # mostramos las primeras 20 para no saturar la salida


Duplicados completos en Ventas: 70 filas


,id_venta,producto,precio,cantidad,fecha_venta,categoria
820,56,Cortinas,$66.24,5.00,05/04/2024,Decoración
821,421,Lámpara de mesa,$114.83,9.00,05/04/2024,Decoración
822,424,Jarrón decorativo,$87.94,2.00,05/04/2024,Decoración
823,1868,Cafetera,$62.23,1.00,05/04/2024,Electrodomésticos
824,2545,Auriculares,$32.81,11.00,05/04/2024,Electrónica
825,2778,SmartWatch,$70.66,7.00,05/04/2024,Electrónica
826,56,Cortinas,$66.24,5.00,05/04/2024,Decoración
827,421,Lámpara de mesa,$114.83,9.00,05/04/2024,Decoración
828,424,Jarrón decorativo,$87.94,2.00,05/04/2024,Decoración
829,1868,Cafetera,$62.23,1.00,05/04/2024,Electrodomésticos


In [21]:
# 7.4.3 — Duplicados completos en Marketing
duplicados_marketing_df = df_marketing[df_marketing.duplicated(keep=False)]
print(f'Duplicados completos en Marketing: {len(duplicados_marketing_df)} filas')
duplicados_marketing_df


Duplicados completos en Marketing: 0 filas


,id_campanha,producto,canal,costo,fecha_inicio,fecha_fin


### 7.5 — Tabla resumen de calidad (los 3 datasets juntos)

Una vista consolidada que facilita comparar el estado de cada dataset.

In [22]:
# 7.5 — Tabla resumen comparativa
resumen_calidad = pd.DataFrame([
    {'dataset': 'clientes',  'registros': len(df_clientes),
     'nulos_total': int(df_clientes.isnull().sum().sum()),
     'duplicados': int(df_clientes.duplicated().sum())},
    {'dataset': 'ventas',    'registros': len(df_ventas),
     'nulos_total': int(df_ventas.isnull().sum().sum()),
     'duplicados': int(df_ventas.duplicated().sum())},
    {'dataset': 'marketing', 'registros': len(df_marketing),
     'nulos_total': int(df_marketing.isnull().sum().sum()),
     'duplicados': int(df_marketing.duplicated().sum())},
])
resumen_calidad['% nulos']      = (resumen_calidad['nulos_total'] / resumen_calidad['registros'] * 100).round(2)
resumen_calidad['% duplicados'] = (resumen_calidad['duplicados']  / resumen_calidad['registros'] * 100).round(2)
resumen_calidad


,dataset,registros,nulos_total,duplicados,% nulos,% duplicados
0,clientes,567,0,0,0.00,0.00
1,ventas,3035,4,35,0.13,1.15
2,marketing,90,0,0,0.00,0.00


## 8. Conclusiones de la Etapa 1

### Estructura general de los datos

| Dataset | Registros | Columnas | Período | Calidad inicial |
|---|---|---|---|---|
| **clientes** | 567 | 5 | - | Sin nulos, sin duplicados |
| **ventas** | 3.035 | 6 | ene-dic 2024 | 4 nulos (`precio`, `cantidad`), 35 duplicados completos |
| **marketing** | 90 | 6 | mar-oct 2024 | Sin nulos, sin duplicados |

### Hallazgos relevantes

**1. Tipos de datos inadecuados en `ventas`:**
- `precio` está como string por el símbolo `$` (ej. `'$69.94'`). Habrá que limpiarlo y convertirlo a `float`.
- `fecha_venta` está como string en formato `dd/mm/yyyy`. Habrá que convertirla a `datetime`.

**2. Tipos de datos inadecuados en `marketing`:**
- Las columnas `fecha_inicio` y `fecha_fin` también están como string. Misma corrección requerida.

**3. Calidad de `ventas`:**
- 35 filas completamente duplicadas (1.15 % del dataset). Cantidad baja pero significativa: hay que eliminarlas para no inflar las métricas de venta.
- 4 valores nulos repartidos entre `precio` (2) y `cantidad` (2). Volumen muy bajo (<0.1 %): se podrían imputar o eliminar las filas afectadas con impacto despreciable.

**4. Vinculación entre datasets:**
- `ventas` y `marketing` se vinculan por la columna `producto` (los 30 productos coinciden 100 % entre ambos). Esto va a permitir analizar la efectividad de las campañas en la Etapa 2.
- `clientes` no se vincula directamente con `ventas` (no hay `id_cliente` en `ventas`). Este dataset servirá para análisis demográfico independiente.

### Próximos pasos (Etapa 2)

1. **Limpieza:**
   - Sacar el `$` de `precio` y convertirlo a `float`.
   - Convertir las 3 columnas de fecha a `datetime`.
   - Eliminar las 35 filas duplicadas en `ventas` (manteniendo la primera ocurrencia).
   - Resolver los 4 nulos de `ventas`: dada la baja cantidad, lo razonable es eliminar esas filas.

2. **Transformación:**
   - Crear columnas derivadas: `importe = precio x cantidad`, `mes`, `año`.
   - Filtrar productos de alto rendimiento.

3. **Agregación:**
   - Resumen de ventas por categoría y por mes.
   - Resumen de inversión en marketing por canal y por producto.

4. **Integración:**
   - Cruzar `ventas` con `marketing` por `producto` para evaluar efectividad de campañas (ROI por canal).


## 9. Anexo

Este apartado describe los archivos que componen la entrega. Todos están alojados en la carpeta de Google Drive del proyecto.

### Carpeta del proyecto

- **Nombre:** `Muñiz Jose - Comisión 26112 - TPI Data Analytics`
- **Link:** [TPI DRIVE](https://drive.google.com/drive/folders/1p642KiplK0Vlv8PFFwoKMqJG4-f7oK-9?usp=sharing)
- **Permisos:** lectura pública habilitada; edición habilitada para el instructor.

### Archivos

| Archivo | Descripción | Link |
|---|---|---|
| `data/clientes.csv` | Dataset original de clientes (567 registros). Sin modificar. | _link_ |
| `data/ventas.csv` | Dataset original de ventas (3.035 registros). Sin modificar. | _link_ |
| `data/marketing.csv` | Dataset original de campañas de marketing (90 registros). Sin modificar. | _link_ |
| `notebook/TPI_DataAnalytics_Etapa1.ipynb` | Este notebook. Desarrollo completo de la Etapa 1. | _link_ |

### Instrucciones de reproducción

El notebook fue desarrollado y ejecutado íntegramente en Google Colaboratory. Para reproducir los resultados:

1. Abrir el notebook en Colab.
2. Ejecutar la celda de montaje de Drive (sección 2.2) y autorizar cuando se solicite.
3. Verificar que la constante `RUTA_PROYECTO` (sección 2.3) apunte a la carpeta correcta.
4. Ejecutar **Runtime → Restart and run all** para correr el notebook completo de principio a fin.

### Stack utilizado

- Python 3
- Pandas
- NumPy
- Google Colaboratory (entorno de ejecución)
- Google Drive (almacenamiento de datasets)


13.1 Estadísticas descriptivas del importe

In [24]:
print("=" * 70)
print("ESTADÍSTICA DESCRIPTIVA - IMPORTE")
print("=" * 70)

print(ventas_calc["importe"].describe())

ESTADÍSTICA DESCRIPTIVA - IMPORTE
count   3,033.00
mean      488.97
std       334.59
min        26.30
25%       220.50
50%       418.05
75%       708.89
max     1,488.12
Name: importe, dtype: float64
